### PR TITLE
Vinai/add drop pks

### PR DESCRIPTION
**Pull request by VinaiRachakonda**

### CODE REVIEW

File: sql/core.go [hunk 1]
```diff
@@ -345,6 +346,17 @@ type CheckAlterableTable interface {
...
+type PrimaryKeyAlterableTable interface {
+	Table
+	// CreatePrimaryKey creates a primary key for this table, using the provided parameters.
+	// Returns an error if the new primary key set is not compatible with the current table data
+	// or the table already has a key.
+	CreatePrimaryKey(ctx *Context, columns []string) error
+	// DropPrimaryKey drops a primary key on a table. Returns an error if that table does not have a key.
+	DropPrimaryKey(ctx *Context) error
+}
+
```
Comment (andrew-wm-arthur) [inline sql/core.go:359]: This interface seems to collide somewhat with `AlterableTable`

Comment (zachmu) [reply sql/core.go:359]: I think not in this case. `AlterableTable` only has ModifyColumn, which changes the definition of a single column.

If anything it kind of overlaps with `IndexAlterableTable`. Adding a primary key is a special case of adding a unique index (the first one for that table). This is actually exactly how MySQL handles that -- the first unique index added to a table without a primary key becomes the primary key.

I think I'm fine with this for now, if for no other reason than we don't support named primary keys, which you need to make this work with `IndexAlterableTable`

File: sql/core.go [hunk 2]
```diff
@@ -345,6 +346,17 @@ type CheckAlterableTable interface {
 	DropCheck(ctx *Context, chName string) error
 }
 
+// PrimaryKeyAlterableTable represents a table that supports primary key changes.
+type PrimaryKeyAlterableTable interface {
+	Table
+	// CreatePrimaryKey creates a primary key for this table, using the provided parameters.
+	// Returns an error if the new primary key set is not compatible with the current table data
+	// or the table already has a key.
```
Comment (zachmu) [inline sql/core.go:354]: Engine should check for the presence of a primary key

File: sql/core.go [hunk 3]
```diff
@@ -345,6 +346,17 @@ type CheckAlterableTable interface {
...
+	CreatePrimaryKey(ctx *Context, columns []string) error
```
Comment (zachmu) [inline sql/core.go:355]: Copy the interface for `IndexAlterableTable` here, but with no indexName param:

`CreateIndex(ctx *Context, indexName string, using IndexUsing, constraint IndexConstraint, columns []IndexColumn, comment string) error`

File: sql/errors.go [hunk 4]
```diff
@@ -293,6 +293,23 @@ var (
 	// ErrTableCopyingNotSupported is returned when a table invokes the TableCopierDatabase interface's
 	// CopyTableData method without supporting the interface
 	ErrTableCopyingNotSupported = errors.NewKind("error: Table copying not supported")
+
+	// ErrMultiplePrimaryKeyDefined is returned when a table invokes CreatePrimaryKey with a primary key already
+	// defined.
+	ErrMultiplePrimaryKeyDefined = errors.NewKind("error: Multiple primary key defined")
```
Comment (zachmu) [inline sql/errors.go:299]: keys

File: sql/errors.go [hunk 5]
```diff
@@ -293,6 +293,23 @@ var (
...
+
+	// ErrWrongAutoKey is returned when a table invokes DropPrimaryKey without first removing the auto increment property
+	// (if it exists) on it.
+	ErrWrongAutoKey = errors.NewKind("error: incorrect table definition; there can be only one auto column and it must be defined as a key")
```
Comment (zachmu) [inline sql/errors.go:303]: : not ;

File: enginetest/enginetests.go [hunk 6]
```diff
@@ -2946,6 +2946,107 @@ func TestShowTableStatus(t *testing.T, harness Harness) {
...
+		// Add back a new primary key and assert the table is queryable
+		TestQuery(t, harness, e, `ALTER TABLE t1 ADD PRIMARY KEY (pk, v)`, []sql.Row{}, nil, nil)
+		TestQuery(t, harness, e, `SELECT * FROM t1`, []sql.Row{
+			{"a1", "a2"},
+			{"a2", "a3"},
+			{"a3", "a4"},
+		}, nil, nil)
+
+		// Drop the original Pk, create an index, create a new primary key
+		TestQuery(t, harness, e, `ALTER TABLE t1 DROP PRIMARY KEY`, []sql.Row{}, nil, nil)
```
Comment (zachmu) [inline enginetest/enginetests.go:3005]: For queries that you aren't making any assertions on, you can just use RunQuery, not TestQuery

**Review by zachmu — COMMENTED** [review]
Mostly looks on the right track, some comments on the interfaces

File: sql/core.go (hunk 3 above)
Comment (VinaiRachakonda) [reply sql/core.go:355]: Does mysql even enforce names primary key constraints? When I was testing the given constraint name was never used in the information_schema.table_constraints name. Seems like this is just adding additional params that will never need to be used.

File: sql/core.go (hunk 1 above)
Comment (VinaiRachakonda) [reply sql/core.go:359]: Created a new interface on the side of caution. Not opinionated on either direction. Will keep this for now.

File: sql/core.go (hunk 3 above)
Comment (VinaiRachakonda) [reply sql/core.go:354]: After discussing it seems like IndexColumn is not enough. indexName, using, and index constraint are not needed

File: enginetest/enginetests.go [hunk 7]
```diff
@@ -2946,6 +2946,107 @@ func TestShowTableStatus(t *testing.T, harness Harness) {
...
+		TestQuery(t, harness, e, `SELECT * FROM t1 ORDER BY pk`, []sql.Row{
+			{"a1", "a2"},
+			{"a1", "a2"},
+			{"a2", "a3"},
+			{"a3", "a4"},
+			{"a4", "a3"},
+		}, nil, nil)
+
+		// Assert that a duplicate row causes an alter table error
+		AssertErr(t, e, harness, `ALTER TABLE t1 ADD PRIMARY KEY (pk, v)`, sql.ErrPrimaryKeyViolation)
```
Comment (zachmu) [inline enginetest/enginetests.go:3046]: Should also check that the schema is unchanged after this

File: memory/table.go [hunk 8]
```diff
@@ -1262,6 +1263,111 @@ Top:
...
+	}
+
+	t.schema = potentialSchema
+	t.partitions = newTable.partitions
+	t.keys = newTable.keys
+
+	return nil
+}
+
+func insertIntoNewTable(t *Table, newSch sql.Schema) (*Table, error) {
```
Comment (zachmu) [inline memory/table.go:1303]: copyTable

**Review by zachmu — APPROVED** [review]
LGTM

**Pull request merged**